src/main/scala/workflow/graph/NodeOptimizationRule.scala

```diff
@@ -0,0 +1,200 @@
+package workflow.graph
+
+import org.apache.spark.rdd.RDD
+import workflow.WorkflowUtils
+
+/**
+ * A modified graph executor used by the NodeOptimizationRule, to keep track
```

Comment:
Also, how different is it from the normal GraphExecutor? Looks like there's probably a fair bit of duplicated logic here?